src/librustc_mir/transform/check_consts/qualifs.rs

```diff
@@ -5,26 +5,67 @@ use rustc::mir::interpret::ConstValue;
 use rustc::ty::{self, Ty};
 use syntax_pos::DUMMY_SP;
 
+use crate::transform::promote_consts::Unpromotable;
 use super::{ConstKind, Item as ConstCx};
 
-#[derive(Clone, Copy)]
-pub struct QualifSet(u8);
+#[derive(Clone, Copy, Default)]
+pub struct QualifSet(pub u8);
 
 impl QualifSet {
-    fn contains<Q: ?Sized + Qualif>(self) -> bool {
+    pub const UNPROMOTABLE: Self = QualifSet(1 << Unpromotable::IDX);
```

Comment:
Can we stop using this hack, btw?

Reply:
There's only two places it can appear now. One is where you left a comment below, which doesn't seem to break anything. The other is when a `const` panics unconditionally, which is just a hack to approximate the existing behavior. I'll try removing it completely.

Reply:
I've removed `UNPROMOTABLE`. Let's see what happens with CI.

Reply:
We pass all the tests, and I've tried the following example locally and no ICE occurs, just an error pointing to `X` saying that const evaluation failed.

```rust
#![feature(const_panic)]
#![allow(const_err)]

const X: u32 = panic!();

fn main() {
    let x: &'static _ = &X;
}
```